clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1001,7 +1005,9 @@ public void commitSync(Duration timeout) {
      * (in which case a {@link org.apache.kafka.common.errors.TimeoutException} is thrown to the caller).
      * <p>
      * Note that asynchronous offset commits sent previously with the {@link #commitAsync(OffsetCommitCallback)}
-     * (or similar) are guaranteed to have their callbacks invoked prior to completion of this method.
+     * (or similar) are guaranteed to have their callbacks invoked prior to completion of this method,
+     * but only when the consumer is using the consumer group protocol.
+     * This guarantee does not hold when using the classic group protocol.
```

Comment:
`but only when the consumer is using the consumer group protocol.` This says it all :)